### PR TITLE
Exchange Redeem Feature

### PR DESCRIPTION
**Pull request by felix2feng**
- Wrote exchangeRedeem feature with unit tests
- Renamed `exchangeIssueParams` to `exchangeInteractData`; Bumped to v30 set-protocol-utils
- Added `depositModule` and `withdrawModule` as convenience methods.
- Added coverage by testing `ExchangeInteractLibrary`

### CODE REVIEW

File: contracts/core/modules/ExchangeRedeemModule.sol [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+     */
+    function exchangeRedeem(
+        ExchangeIssueLibrary.ExchangeIssueParams memory _exchangeInteractData,
+        bytes memory _orderData
+    )
+        public
+        nonReentrant
+    {
+        // Validate ExchangeRedeemParams
+        validateExchangeIssueParams(_exchangeInteractData);
```
Comment (bweick) [inline contracts/core/modules/ExchangeRedeemModule.sol]: Can we give this a more general name? The juxtapostion between the comment and this is kinda confusing

Comment (felix2feng) [reply contracts/core/modules/ExchangeRedeemModule.sol]: Yeah in a separate PR, I'm going to rename everything to `exchangeIssuance` (requires set-protocol-utils updates)

Comment (felix2feng) [reply contracts/core/modules/ExchangeRedeemModule.sol]: I ended up adding it to this PR

File: contracts/core/lib/ExchangeInteractLibrary.sol [hunk 2]
```diff
@@ -22,20 +22,30 @@ import { IVault } from "../interfaces/IVault.sol";
...
+library ExchangeInteractLibrary {
+    // ============ Structs ============
+    struct ExchangeInteractData {
+        address setAddress;
+        uint256 quantity;
+        uint8[] sentTokenExchangeIds;
+        address[] sentTokens;
+        uint256[] sentTokenAmounts;
+        address[] receiveTokens;
+        uint256[] receiveTokenAmounts;
```
Comment (asoong) [inline contracts/core/lib/ExchangeInteractLibrary.sol]: One is past tense and the other is present. in favor of `receivedTokens`

File: test/contracts/core/modules/exchangeRedeemModule.spec.ts [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+  let exchangeRedeemModule: ExchangeRedeemModuleContract;
+  let setTokenFactory: SetTokenFactoryContract;
+
+  const coreWrapper = new CoreWrapper(contractDeployer, contractDeployer);
+  const erc20Wrapper = new ERC20Wrapper(contractDeployer);
+  const exchangeWrapper = new ExchangeWrapper(contractDeployer);
+
+  before(async () => {
+    ABIDecoder.addABI(Core.abi);
+    ABIDecoder.addABI(ExchangeRedeemModule.abi);
```
Comment (asoong) [inline test/contracts/core/modules/exchangeRedeemModule.spec.ts]: Can we add multiple ABIs? I cant find it but I swear in the past there was a rule that we could only add one.

File: contracts/core/interfaces/ICore.sol [hunk 4]
```diff
@@ -285,15 +285,15 @@ interface ICore {
      * Exposes internal function that withdraws a quantity of tokens from the vault and
      * deattributes the tokens respectively, to system modules.
      *
-     * @param  _token           Address of token being withdrawn
      * @param  _from            Address to decredit for withdraw
      * @param  _to              Address to transfer tokens to
+     * @param  _token           Address of token being withdrawn
```
[the remark this answers is not in the document]
Comment (felix2feng) [reply contracts/core/interfaces/ICore.sol]: Ok - I'm happy either way as long as its consistent. I think it got muddled when we started adding the batch functions where the arrays are put on the bottom

```function batchDepositModule(
        address _from,
        address _to,
        address[] calldata _tokens,
        uint256[] calldata _quantities
    )
```

[the remark this answers is not in the document]
Comment (felix2feng) [reply contracts/core/interfaces/ICore.sol]: Can decide once and for all in a subsequent PR

File: contracts/core/lib/ExchangeInteractLibrary.sol [hunk 5]
```diff
@@ -22,20 +22,30 @@ import { IVault } from "../interfaces/IVault.sol";
 
 
 /**
- * @title ExchangeValidationLibrary
+ * @title ExchangeInteractLibrary
```
[the remark this answers is not in the document]
Comment (felix2feng) [reply contracts/core/lib/ExchangeInteractLibrary.sol]: open for other names. Other idea is `ExchangeIssuanceLibrary` (hopefully implying issuance covers issue and redeem)

File: contracts/core/modules/ExchangeRedeemModule.sol [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+import { ExchangeExecution } from "./lib/ExchangeExecution.sol";
+import { ExchangeInteractLibrary } from "../lib/ExchangeInteractLibrary.sol";
+import { ISetToken } from "../interfaces/ISetToken.sol";
+import { ModuleCoreState } from "./lib/ModuleCoreState.sol";
+
+
+/**
+ * @title ExchangeRedeemModule
```
[the remark this answers is not in the document]
Comment (felix2feng) [reply contracts/core/modules/ExchangeRedeemModule.sol]: +1

File: test/contracts/core/modules/exchangeRedeemModule.spec.ts (hunk 3 above)
Comment (felix2feng) [reply test/contracts/core/modules/exchangeRedeemModule.spec.ts]: Yeah i believe so. Creation and Checking exchange redeem logs are working

File: contracts/core/extensions/CoreModuleInteraction.sol [hunk 7]
```diff
@@ -49,7 +49,42 @@ contract CoreModuleInteraction is
...
+    function depositModule(
+        address _from,
+        address _to,
+        address _token,
+        uint256 _quantity
+    )
+        external
+        onlyModule
+    {
+
```
Comment (asoong) [inline contracts/core/extensions/CoreModuleInteraction.sol]: whitespace

File: contracts/core/extensions/CoreModuleInteraction.sol [hunk 8]
```diff
@@ -49,7 +49,42 @@ contract CoreModuleInteraction is
...
+        onlyModule
+    {
+        
+        address[] memory tokenArray = new address[](1);
+        tokenArray[0] = _token;
+
+        uint256[] memory quantityArray = new uint256[](1);
+        quantityArray[0] = _quantity;
+        
+
```
Comment (asoong) [inline contracts/core/extensions/CoreModuleInteraction.sol]: whitespace

File: contracts/core/extensions/CoreModuleInteraction.sol [hunk 9]
```diff
@@ -75,6 +110,41 @@ contract CoreModuleInteraction is
...
+    function withdrawModule(
+        address _from,
+        address _to,
+        address _token,
+        uint256 _quantity
+    )
+        external
+        onlyModule
+    {
+
```
Comment (asoong) [inline contracts/core/extensions/CoreModuleInteraction.sol]: whitespace

File: contracts/core/extensions/CoreModuleInteraction.sol [hunk 10]
```diff
@@ -75,6 +110,41 @@ contract CoreModuleInteraction is
...
+        onlyModule
+    {
+        
+        address[] memory tokenArray = new address[](1);
+        tokenArray[0] = _token;
+
+        uint256[] memory quantityArray = new uint256[](1);
+        quantityArray[0] = _quantity;
+        
+
```
Comment (asoong) [inline contracts/core/extensions/CoreModuleInteraction.sol]: whitespace

File: contracts/core/lib/ExchangeIssuanceLibrary.sol [hunk 11]
```diff
@@ -127,45 +134,46 @@ library ExchangeValidationLibrary {
...
         internal
         view
     {
         require(
-            _sentTokenExchanges.length == _sentTokens.length && 
-            _sentTokens.length == _sentTokenAmounts.length,
-            "ExchangeValidationLibrary.validateSentTokenParams: Sent token inputs must be of the same length"
+            _sendTokenExchangeIds.length == _sendTokens.length && 
+            _sendTokens.length == _sendTokenAmounts.length,
+            "ExchangeIssuanceLibrary.validateSendTokenParams: Sent token inputs must be of the same length"
```
Comment (asoong) [inline contracts/core/lib/ExchangeIssuanceLibrary.sol]: `Send`?

**Pull request merged**